### PR TITLE
plugin.video.themoviedb.helper 2.0.18

### PR DESCRIPTION
**Pull request by jurialmunkey**
### Description

This update adds several significant updates and improvements. Key new features include:

- A listitem service monitor that pulls additional details from tmdb and omdb.
- A context menu for trakt watchlist/collection/history add/remove functions.
- Icons for categories.
- Various new plugin lists, including a randomised list section for usage as widgets. 
- A player function that allows users to play/search other addons for content.
- User options to obtain additional artwork and metadata from fanarttv and local db.
- A trakt calendar option for next aired and recently aired tvshows

### Checklist:

- [x] My code follows the [add-on rules](http://kodi.wiki/view/Add-on_rules) and [piracy stance](http://kodi.wiki/view/Official:Forum_rules#Piracy_Policy) of this project. 
- [x] I have read the [CONTRIBUTING](https://github.com/xbmc/repo-plugins/blob/master/CONTRIBUTING.md) document
- [x] Each add-on submission should be a single commit with using the following style: [plugin.video.foo] v1.0.0

Additional information :
- Submitting your add-on to this specific branch makes it available to any Kodi version equal or higher than the branch name with the applicable Kodi dependencies limits.
- [add-on development](http://kodi.wiki/view/Add-on_development) wiki page.
- Kodi [pydocs](http://kodi.wiki/view/PyDocs) provide information about the Python API
- [PEP8](https://www.python.org/dev/peps/pep-0008/) codingstyle which is considered best practice but not mandatory.
- This add-on repository has automated code guideline check which could help you improve your coding. You can find the results of these check at [Codacy](https://www.codacy.com/app/Kodi/repo-plugins/dashboard). You can create your own account as well to continuously monitor your python coding before submitting to repo.
- Development questions can be asked in the [add-on development](http://forum.kodi.tv/forumdisplay.php?fid=26) section on the Kodi forum.
- If you see no activity on your PR after a week (so at least one weekend has passed) then please go to the #kodi-dev freenode IRC channel to reach out to the team

### CODE REVIEW

**Comment by jurialmunkey** [comment]
Added latest commit for a feature that should make it into this release (also fixed the Travis build error).

File: plugin.video.themoviedb.helper/resources/lib/container.py [hunk 1]
```diff
@@ -297,97 +263,316 @@ def list_traktuserlists(self):
...
+        d_args = (
+            ('tv', self.params.get('tmdb_id'), self.params.get('season'), self.params.get('episode'))
+            if self.params.get('type') == 'episode' else (self.params.get('type'), self.params.get('tmdb_id')))
+
+        # Check if we want to refresh cache with &amp;refresh=True
+        if self.params.get('refresh') == 'True':
+            with utils.busy_dialog():
+                self.tmdb.get_detailed_item(*d_args, cache_refresh=True)
+            xbmc.executebuiltin('Container.Refresh')
+            xbmcgui.Dialog().ok('Cache Refresh', 'Cached details were refreshed')
```
Comment (enen92) [inline plugin.video.themoviedb.helper/resources/lib/container.py]: Strings presented to the user in the interface need to be localized

File: plugin.video.themoviedb.helper/resources/lib/container.py [hunk 2]
```diff
@@ -297,97 +263,316 @@ def list_traktuserlists(self):
...
+            return
+        self.plugincategory = recentitem[2]
+        self.params['tmdb_id'] = recentitem[1]
+        self.params['info'] = 'recommendations'
+        self.list_tmdb()
+
+    def list_getid(self):
+        self.params['tmdb_id'] = self.get_tmdb_id(**self.params)
+
+    def list_play(self):
```
Comment (enen92) [inline plugin.video.themoviedb.helper/resources/lib/container.py]: I assume this is only called when playing a playlist? Our api should be improved in this regard

File: plugin.video.themoviedb.helper/resources/lib/container.py [hunk 3]
```diff
@@ -297,97 +263,316 @@ def list_traktuserlists(self):
...
+                    item.poster = item.icon = i.get('icon', '').format(self.addonpath)
+                items.append(item)
+
+        # Add trakt management items if &amp;manage=True
+        if self.addon.getSettingBool('trakt_management'):
+            traktapi = traktAPI()
+            trakt_collection = traktapi.sync_collection(utils.type_convert(self.params.get('type'), 'trakt'), 'tmdb')
+            if trakt_collection:
+                boolean = 'remove' if details.get('tmdb_id') in trakt_collection else 'add'
+                item_collection = ListItem(library=self.library, **details)
```
Comment (enen92) [inline plugin.video.themoviedb.helper/resources/lib/container.py]: Localize

File: plugin.video.themoviedb.helper/resources/lib/container.py [hunk 4]
```diff
@@ -297,97 +263,316 @@ def list_traktuserlists(self):
...
+                item_collection.label = 'Remove from Trakt Collection' if boolean == 'remove' else 'Add to Trakt Collection'
+                item_collection.url = {'info': 'details', 'trakt': 'collection_{0}'.format(boolean), 'type': self.params.get('type')}
+                items.append(item_collection)
+            trakt_watchlist = traktapi.sync_watchlist(utils.type_convert(self.params.get('type'), 'trakt'), 'tmdb')
+            if trakt_watchlist:
+                boolean = 'remove' if details.get('tmdb_id') in trakt_watchlist else 'add'
+                item_watchlist = ListItem(library=self.library, **details)
+                item_watchlist.label = 'Remove from Trakt Watchlist' if boolean == 'remove' else 'Add to Trakt Watchlist'
```
Comment (enen92) [inline plugin.video.themoviedb.helper/resources/lib/container.py]: Same

File: plugin.video.themoviedb.helper/resources/lib/container.py [hunk 5]
```diff
@@ -297,97 +263,316 @@ def list_traktuserlists(self):
...
+                item_watchlist.url = {'info': 'details', 'trakt': 'watchlist_{0}'.format(boolean), 'type': self.params.get('type')}
+                items.append(item_watchlist)
+            trakt_history = traktapi.sync_history(utils.type_convert(self.params.get('type'), 'trakt'), 'tmdb')
+            if trakt_history:
+                boolean = 'remove' if details.get('tmdb_id') in trakt_history else 'add'
+                item_history = ListItem(library=self.library, **details)
+                item_history.label = 'Remove from Trakt Watched History' if boolean == 'remove' else 'Add to Trakt Watched History'
```
Comment (enen92) [inline plugin.video.themoviedb.helper/resources/lib/container.py]: Same here pls

File: plugin.video.themoviedb.helper/resources/lib/listitem.py [hunk 6]
```diff
@@ -1,46 +1,215 @@
...
-        self.icon = icon if icon else '{0}/resources/poster.png'.format(_addonpath)
-        self.fanart = fanart if fanart else '{0}/fanart.jpg'.format(_addonpath)
-        self.cast = cast if cast else []  # Cast list
+
+
+class ListItem(object):
+    def __init__(self, label=None, label2=None, dbtype=None, library=None, tmdb_id=None, imdb_id=None, dbid=None, tvdb_id=None,
+                 cast=None, infolabels=None, infoproperties=None, poster=None, thumb=None, icon=None, fanart=None, nextpage=None,
+                 streamdetails=None, clearlogo=None, clearart=None, banner=None, landscape=None, mixed_type=None, url=None, is_folder=True):
+        self.addon = xbmcaddon.Addon('plugin.video.themoviedb.helper')
```
Comment (enen92) [inline plugin.video.themoviedb.helper/resources/lib/listitem.py]: Id not needed here. Addon() is enough

File: plugin.video.themoviedb.helper/resources/lib/container.py (hunk 2 above)
Comment (jurialmunkey) [reply plugin.video.themoviedb.helper/resources/lib/container.py]: It's a badly named function - it's a call back play function for listitems. The Player module takes tmdb_id, looks-up details online and the compares against local db. If a matching item is found in local db, it is played. 

If a local item can't be found, plugin presents user with options to play or search with an external plugin. The actions to play with an external plugin are defined in player.json files which are placed in `userdata/addon_data/plugin.video.themoviedb.helper/players/`

**Pull request merged**